### PR TITLE
Credits and Inventory

### PR DESCRIPTION
**Pull request by o-psi**
Credits, and Inventory modules.

### CODE REVIEW

**Comment by wrongecho** [comment]
Test these changes at: https://0185905.pr-review.itflow.org  
 _(automatic message)_

**Comment by wrongecho** [comment]
Unsure why #897 was closed, but would recommend credits and inventory are raised as separate PRs to allow for proper testing.

**Comment by johnnyq** [comment]
This is fine for now just for the future one PR per feature will make it easier to manage and review. I see this is still is draft so will review when ready, looking forward to credits and Inventory.

**Comment by johnnyq** [comment]
This looks pretty good to me,
@o-psi has it been working good for you guys over there
@wrongecho how does it look to you

**Comment by wrongecho** [comment]
On initial review, I'm unable to get the PR Review database past 1.1.0.

File: credits.php [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+                        $credit_amount = $row['credit_amount'];
+                        $credit_currency_code = $row['credit_currency_code'];
+                        $credit_date = $row['credit_date'];
+                        $credit_reference = $row['credit_reference'];
+                        $credit_client_id = $row['credit_client_id'];
+                        $credit_payment_id = $row['credit_payment_id'];
+                        $credit_account_id = $row['credit_account_id'];
+
+                        // Get client name from DB
+                        $clientQuery = mysqli_query($mysqli, "SELECT * FROM clients WHERE client_id = $credit_client_id");
```
Comment (wrongecho) [inline credits.php:134]: To save on database calls, can the client name be pulled from the earlier $sql join query at the top of the page?

File: db.sql [hunk 2]
```diff
@@ -1505,6 +1606,20 @@ CREATE TABLE `ticket_attachments` (
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_general_ci;
 /*!40101 SET character_set_client = @saved_cs_client */;
 
+--
+-- Table structure for table `ticket_products`
```
Comment (wrongecho) [inline db.sql:1610]: Ticket products seems to be a new table defined in db.sql, but seems to not be added during the database upgrade?

File: post/inventory_model.php [hunk 3]
```diff
@@ -0,0 +1 @@
+<?php
```
Comment (wrongecho) [inline post/inventory_model.php:1]: ?

File: ticket_add_product_modal.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                    <button type="button" class="btn btn-light" data-dismiss="modal"><i class="fa fa-times mr-2"></i>Cancel</button>
+                    <!-- Merge button starts disabled. Is enabled by the merge_into_number_get_details Javascript function-->
+                </div>
+            </form>
+        </div>
+    </div>
+</div>
+
+<!-- Ticket merge JS -->
+<script src="js/ticket_merge.js"></script>
```
Comment (wrongecho) [inline ticket_add_product_modal.php:46]: Likely copied by mistake, can probably be removed?

**Comment by wrongecho** [comment]
@johnnyq Have added a few comments but on mobile atm. Will try and take a look this week once the update queries are working.

**Review by wrongecho — CHANGES_REQUESTED** [review]

**Comment by o-psi** [comment]
@johnnyq When you invited me to fork the repository, that is when we started doing version numbers. 
And it is the fact that I am doing one PR per version number, rather than feature. This is as, everything from 0.1.8.4 would have been in this PR regardless.

File: credits.php (hunk 1 above)
Comment (o-psi) [reply credits.php:134]: Definitely can. I used a join statement with those intentions. Many thanks for the catch.

File: db.sql (hunk 2 above)
Comment (o-psi) [reply db.sql:1610]: Good catch. This was resolved.

File: ticket_add_product_modal.php (hunk 4 above)
Comment (o-psi) [reply ticket_add_product_modal.php:46]: Indeed

File: post/inventory_model.php (hunk 3 above)
Comment (o-psi) [reply post/inventory_model.php:1]: This was to remind me to implement the post in a model. Good catch!

**Comment by o-psi** [comment]
I cannot mark the changes as accepted, but all comments have been resolved. We have now been using for a few days, and have yet to find any quirks.

File: credits.php [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+                    while ($row = mysqli_fetch_array($sql)) {
+                        $credit_id = $row['credit_id'];
+                        $credit_amount = $row['credit_amount'];
+                        $credit_currency_code = $row['credit_currency_code'];
+                        $credit_date = $row['credit_date'];
+                        $credit_reference = $row['credit_reference'];
+                        $credit_client_id = $row['credit_client_id'];
+                        $credit_payment_id = $row['credit_payment_id'];
+                        $credit_account_id = $row['credit_account_id'];
+                        $client_name = $row['client_name'];
```
Comment (johnnyq) [inline credits.php:132]: @o-psi These will have be sanitized for output

File: credits.php [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+                        // Get account name from DB
+                        if($credit_account_id != NULL) {
+                            $accountQuery = mysqli_query($mysqli, "SELECT * FROM accounts WHERE account_id = $credit_account_id");
+                            $account = mysqli_fetch_array($accountQuery);
+                            $account_name = $account['account_name'];
```
Comment (johnnyq) [inline credits.php:138]: same with this one

File: credits.php [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+                        } else {
+                            $account_name = "Unassigned";
+                        }
+
+                        // Get payment invoice and reference from DB
+                        if($credit_payment_id != NULL) {
+                            $paymentQuery = mysqli_query($mysqli, "SELECT * FROM payments WHERE payment_id = $credit_payment_id");
+                            $payment = mysqli_fetch_array($paymentQuery);
+                            $payment_invoice = $payment['payment_invoice_id'];
+                            $payment_reference = $payment['payment_reference'];
```
Comment (johnnyq) [inline credits.php:148]: sanitize these for output

File: credits.php [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+                            $payment_invoice = "Unassigned";
+                            $payment_reference = "Unassigned";
+                        }
+
+                        // Get invoice prefix and number from DB
+                        if($payment_invoice != NULL) {
+                            $invoiceQuery = mysqli_query($mysqli, "SELECT * FROM invoices WHERE invoice_id = $payment_invoice");
+                            $invoice = mysqli_fetch_array($invoiceQuery);
+                            $invoice_prefix = $invoice['invoice_prefix'];
+                            $invoice_number = $invoice['invoice_number'];
```
Comment (johnnyq) [inline credits.php:159]: sanitize these for output

**Review by johnnyq — CHANGES_REQUESTED** [review]
Please sanitize the outputs of these guys to prevent XSS Attacks

File: database_updates.php [hunk 9]
```diff
@@ -1636,11 +1638,55 @@
...
+
+        mysqli_query($mysqli, "CREATE TABLE `ticket_products` (
+            `ticket_product_id` int(11) NOT NULL AUTO_INCREMENT,
+            `ticket_product_product_id` int(11) NOT NULL,
+            `ticket_product_ticket_id` int(11) NOT NULL,
+            `ticket_product_quantity` int(11) NOT NULL,
+            PRIMARY KEY (`ticket_product_id`)
+          ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_general_ci");
+
+        mysqli_query($mysqli, "CREATE TABLE `credits ` (`credit_id` int(11) NOT NULL AUTO_INCREMENT,`credit_amount` decimal(15,2) NOT NULL,`credit_currency_code` varchar(200) NOT NULL,`credit_date` date NOT NULL,`credit_reference` text DEFAULT NULL,`credit_created_at` datetime NOT NULL DEFAULT current_timestamp(),`credit_updated_at` datetime DEFAULT NULL ON UPDATE current_timestamp(),`credit_archived_at` datetime DEFAULT NULL, `credit_client_id` int(11) NOT NULL,`credit_payment_id` int(11) NOT NULL,`credit_account_id` int(11) NOT NULL, PRIMARY KEY (`credit_id`))");
```
Comment (wrongecho) [inline database_updates.php:1684]: There is an extra space after the word credits preventing the creation of the table during the database update.

**Comment by johnnyq** [comment]
Something still appears to broken on Database updates.

File: post/inventory.php [hunk 10]
```diff
@@ -0,0 +1,97 @@
+<?php
+
+/*
+ * ITFlow - GET/POST request handler for AI Functions
```
Comment (wrongecho) [inline post/inventory.php:4]: "AI Functions" should probably read inventory.

**Review by wrongecho — CHANGES_REQUESTED** [review]
Still struggling to get the update correctly working on my side, syntax issues and mismatch between updated db version and latest db version. 
Also looks as though there is a syntax issue in db.sql which would break new installs.

Edit: it would also be great if we could get that Sonar issue count down. Sometimes it's overzealous, but I don't think I've ever seen so many new issues in one PR. Appreciate you're trying to get the bare bones of this working as quickly as possible, which is fine in your fork, but ideally we're aiming for maintainable code here as many others will be relying on it.

File: ticket.php [hunk 11]
```diff
@@ -881,6 +892,15 @@
                     </div>
                     <!-- End Vendor card -->
 
+                    <!-- Products card -->
```
Comment (wrongecho) [inline ticket.php:895]: In the main side nav (side_nav.php), we only show products if config_module_enable_accounting is 1. I think this should be the same?

File: ticket.php [hunk 12]
```diff
@@ -952,6 +972,8 @@
 
         require_once "ticket_merge_modal.php";
 
+        require_once "ticket_add_product_modal.php";
```
Comment (wrongecho) [inline ticket.php:975]: Same here re config_module_enable_accounting, can we drop this inside the if statement below?

File: post/invoice.php [hunk 13]
```diff
@@ -1428,3 +1453,154 @@
...
+        // Add to Email Body Invoice Portion
+
+        $email_body_invoices .= "<br>Invoice <a href=\'https://$config_base_url/guest_view_invoice.php?invoice_id=$invoice_id&url_key=$invoice_url_key\'>$invoice_prefix$invoice_number</a> - Outstanding Amount: " . numfmt_format_currency($currency_format, $invoice_balance, $currency_code) . " - Payment Applied: " . numfmt_format_currency($currency_format, $payment_amount, $currency_code) . " - New Balance: " . numfmt_format_currency($currency_format, $remaining_invoice_balance, $currency_code);
+
+    } // End Invoice Loop
+
+
+
+    // Send Email
+    if ($email_receipt == 1) {
```
Comment (wrongecho) [inline post/invoice.php:1551]: $email_receipt does not appear to be assigned via post or within the apply_credit logic?

File: post/invoice.php [hunk 14]
```diff
@@ -1428,3 +1453,154 @@
...
+
+    $_SESSION['alert_message'] .= "Credit applied to $credit_applied_count invoices";
+    
+    header("Location: " . $_SERVER["HTTP_REFERER"]);
+}
+
+if (isset($_GET['delete_credit'])) {
+    $credit_id = intval($_GET['delete_credit']);
+
+    mysqli_query($mysqli,"DELETE FROM credits WHERE credit_id = $credit_id");
```
Comment (wrongecho) [inline post/invoice.php:1602]: Needs logging

File: post/invoice.php [hunk 15]
```diff
@@ -610,173 +611,197 @@
...
-        $client_id = intval($row['client_id']);
-        $client_name = sanitizeInput($row['client_name']);
-        $contact_name = sanitizeInput($row['contact_name']);
-        $contact_email = sanitizeInput($row['contact_email']);
-        $contact_phone = sanitizeInput(formatPhoneNumber($row['contact_phone']));
-        $contact_extension = preg_replace("/[^0-9]/", '',$row['contact_extension']);
-        $contact_mobile = sanitizeInput(formatPhoneNumber($row['contact_mobile']));
+    if($payment_is_credit) {
+    //Create a credit for the overpayment
+    mysqli_query($mysqli,"INSERT INTO credits SET credit_amount = $credit_amount, credit_currency_code = '$currency_code', credit_date = '$date', credit_reference = 'Overpayment: $reference', credit_client_id = (SELECT invoice_client_id FROM invoices WHERE invoice_id = $invoice_id), credit_payment_id = $payment_id, credit_account_id = $account");
```
Comment (wrongecho) [inline post/invoice.php:633]: Needs logging

File: inventory_locations_manage.php [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+        LEFT JOIN users on inventory_locations_user_id = user_id
+        LEFT JOIN vendors ON inventory_vendor_id = vendor_id
+        WHERE inventory_location_id = $inventory_location_id
+        GROUP BY inventory_product_id
+        ");
+
+$num_rows = mysqli_fetch_row(mysqli_query($mysqli, "SELECT FOUND_ROWS()"));
+
+$location_name_row = mysqli_fetch_row(mysqli_query($mysqli,"SELECT inventory_locations_name FROM inventory_locations WHERE inventory_locations_id = $inventory_location_id"));
+$location_name = $location_name_row[0];
```
Comment (wrongecho) [inline inventory_locations_manage.php:35]: Believe this should be sanitized

File: db.sql [hunk 17]
```diff
@@ -662,6 +685,54 @@ CREATE TABLE `interfaces` (
...
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_general_ci;
+/*!40101 SET character_set_client = @saved_cs_client */;
+
+--
+-- Table Structure for `inventory_locations`
+--
+DROP TABLE IF EXISTS `inventory_locations`;
+/*!40101 SET @saved_cs_client     = @@character_set_client */;
+/*!40101 SET character_set_client = utf8 */;
+CREATE TABLE `inventory_locations` (
```
Comment (wrongecho) [inline db.sql:719]: Worth checking with @johnnyq but believe the table name is plural (correct) but the column names should be singular

https://docs.itflow.org/code_standards#sql_table_structure
_Column Naming: Define by non-plural word of the table name separated with an underscore. Example table users has user_id user_first_name_

File: inventory.php [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+                            <th><a class="text-dark" href="?<?php echo $url_query_strings_sort; ?>&sort=inventory_date&order=<?php echo $disp; ?>">Product Name</a></th>
+                            <th><a class="text-dark" href="?<?php echo $url_query_strings_sort; ?>&sort=vendor_name&order=<?php echo $disp; ?>">Quantity</a></th>
+                            <th><a class="text-dark" href="?<?php echo $url_query_strings_sort; ?>&sort=category_name&order=<?php echo $disp; ?>">Locations</a></th>
+                            <th class="text-center">Manage product</th>
+                        </tr>
+                        </thead>
+                        <tbody>
+                        <?php
+
+                        while ($row = mysqli_fetch_array($sql)) {
```
Comment (wrongecho) [inline inventory.php:93]: Believe these should be sanitized

**Comment by o-psi** [comment]
There seems to be a few commits missing that must be stashed somewhere. I know I went through and sanitized most if not all of these mentioned. Many thanks for taking the time to recognize that.

File: credits.php (hunk 5 above)
Comment (o-psi) [reply credits.php:132]: Done

File: db.sql [hunk 19]
```diff
@@ -874,6 +945,36 @@ CREATE TABLE `payments` (
...
+  `purchase_order_discount` decimal(15,2) NOT NULL DEFAULT 0.00,
+  PRIMARY KEY (`purchase_order_id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci;
+/*!40101 SET character_set_client = @saved_cs_client */;
+
+--
+-- Table structure for table `purchase_order_items`
+--
+
+DROP TABLE IF EXISTS
```
[the remark this answers is not in the document]
Comment (o-psi) [reply db.sql:974]: This was for another branch. Good catch

**Comment by johnnyq** [comment]
Hey @o-psi the field names should be singular for the table inventory_locations to keep consistency
example inventory_location_id

File: db.sql (hunk 17 above)
Comment (johnnyq) [reply db.sql:719]: yes the field names should be singular good catch

**Comment by o-psi** [comment]
Crazy weekend. This has been fixed.

**Comment by johnnyq** [comment]
> Crazy weekend. This has been fixed.

Yeah man same here
We'll go ahead and review further

**Pull request closed without merging**